metricbeat/module/windows/wmi/utils.go

```diff
@@ -0,0 +1,200 @@
+// method in the wmi.wmiInstance struct and applies it on the Properties_ field
+// https://github.com/microsoft/wmi/blob/v0.25.2/pkg/wmiinstance/WmiInstance.go#L87
+//
+// Note: We are not instantiating a wmi.wmiProperty because of this issue
+// https://github.com/microsoft/wmi/issues/150
+// Once this issue is resolved, we can instantiate a wmi.WmiProperty and eliminate
+// the need for the "getPropertyType" function.
+func getProperty(instance *wmi.WmiInstance, propertyName string, logger *logp.Logger) (*ole.IDispatch, error) {
+	// Documentation: https://learn.microsoft.com/en-us/windows/win32/wmisdk/swbemobject-properties-
+	rawResult, err := oleutil.GetProperty(instance.GetIDispatch(), "Properties_")
+	if err != nil {
+		return nil, err
+	}
+
+	// SWbemObjectEx.Properties_ returns
+	// an SWbemPropertySet object that contains the collection
+	// of properties for the c class
+	sWbemObjectExAsIDispatch := rawResult.ToIDispatch()
+	defer func() {
+		if cerr := rawResult.Clear(); cerr != nil {
+			logger.Debugf("failed to release connection: %w", err)
+		}
+	}()
+
+	// Get the property
+	sWbemProperty, err := oleutil.CallMethod(sWbemObjectExAsIDispatch, "Item", propertyName)
+	if err != nil {
+		return nil, err
+	}
+
+	return sWbemProperty.ToIDispatch(), nil
+}
+
+// Given an instance and a property Name, it returns the appropriate conversion function
+func GetConvertFunction(instance *wmi.WmiInstance, propertyName string, logger *logp.Logger) (WmiStringConversionFunction, error) {
+	rawProperty, err := getProperty(instance, propertyName, logger)
+	if err != nil {
+		return nil, err
+	}
+	propType := getPropertyType(rawProperty)
+
+	var f WmiStringConversionFunction
+
+	switch propType {
+	case base.WbemCimtypeDatetime:
+		f = ConvertDatetime
+	case base.WbemCimtypeUint64:
+		f = ConvertUint64
+	case base.WbemCimtypeSint64:
+		f = ConvertSint64
+	default: // For all other types we return the identity function
+		f = ConvertString
+	}
+	return f, err
+}
+
+// Utilities related to Warning Threshold
+
+// Define an interface to allow unit-testing long-running queries
+// *wmi.wmiSession is an implementation of this interface
+type WmiQueryInterface interface {
+	QueryInstances(query string) ([]*wmi.WmiInstance, error)
+}
+
+// Wrapper of the session.QueryInstances function that execute a query for at most a timeout
+// after which we stop actively waiting.
+// Note that the underlying query will continue to run, until the query completes or the WMI Arbitrator stops the query
+// https://learn.microsoft.com/en-us/troubleshoot/windows-server/system-management-components/new-wmi-arbitrator-behavior-in-windows-server
+func ExecuteGuardedQueryInstances(session WmiQueryInterface, query string, timeout time.Duration, logger *logp.Logger) ([]*wmi.WmiInstance, error) {
+	var rows []*wmi.WmiInstance
+	var err error
+	done := make(chan error)
+	timedout := false
+
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
```

Comment:
this just ensures the context is always cleaned up when the query completes within the timeout?